### PR TITLE
KAFKA-15045: (KIP-924 pt. 15) Implement #defaultStandbyTaskAssignment and finish rack-aware standby optimization

### PR DESCRIPTION
**Pull request by apourchet**
This fills in the implementation details of the standby task assignment utility functions within TaskAssignmentUtils.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java [hunk 1]
```diff
@@ -62,14 +64,24 @@ public static KafkaStreamsAssignment of(final ProcessId processId, final Set<Ass
...
                                    final Set<AssignedTask> assignment,
                                    final Optional<Instant> followupRebalanceDeadline) {
+        this(
+            processId,
+            assignment.stream().collect(Collectors.toMap(AssignedTask::id, Function.identity())),
+            followupRebalanceDeadline
+        );
+    }
+
+    private KafkaStreamsAssignment(final ProcessId processId,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java]: nit: seems kind of unnecessary to introduce yet another constructor for this, we can just inline things. ie

```
this.assignment = assignment.stream().collect(Collectors.toMap(AssignedTask::id, Function.identity()));
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java [hunk 2]
```diff
@@ -83,16 +95,10 @@ public ProcessId processId() {
...
-    public Set<AssignedTask> assignment() {
-        // TODO change assignment to return a map so we aren't forced to copy this into a Set
-        return new HashSet<>(assignment.values());
-    }
-
-    // TODO: merge this with #assignment by having it return a Map<TaskId, AssignedTask>
-    public Set<TaskId> assignedTaskIds() {
-        return assignment.keySet();
+    public Map<TaskId, AssignedTask> tasks() {
+        return unmodifiableMap(assignment);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java]: nit: should probably changed the field name too (ie `assignment` --> `tasks`)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java [hunk 3]
```diff
@@ -278,7 +278,7 @@ public void processOptimizedAssignments(final Map<ProcessId, KafkaStreamsAssignm
 
             for (final Map.Entry<ProcessId, KafkaStreamsAssignment> entry : optimizedAssignments.entrySet()) {
                 final ProcessId processId = entry.getKey();
-                final Set<AssignedTask> assignedTasks = optimizedAssignments.get(processId).assignment();
+                final Set<AssignedTask> assignedTasks = new HashSet<>(optimizedAssignments.get(processId).tasks().values());
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java:281]: This can be a followup PR, but one of the nice things about making KafkaStreamsAssignment mutable is that we should be able to get rid of the `newAssignments` field altogether and stop converting back and forth between the KafkaStreamsAssignment and raw Sets so that we can add/remove tasks. 

In addition to the obvious code simplification that should result from this change, it should save us a lot of copying things into various different data structures and reduce a lot of overhead

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 4]
```diff
@@ -785,12 +789,13 @@ private UserTaskAssignmentListener assignTasksToClients(final Cluster fullMetada
             );
             final org.apache.kafka.streams.processor.assignment.TaskAssignor assignor = userTaskAssignor.get();
             final TaskAssignment taskAssignment = assignor.assign(applicationState);
-            processStreamsPartitionAssignment(clientMetadataMap, taskAssignment);
             final AssignmentError assignmentError = validateTaskAssignment(applicationState, taskAssignment);
-            userTaskAssignmentListener = (GroupAssignment assignment, GroupSubscription subscription) -> {
+            processStreamsPartitionAssignment(clientMetadataMap, taskAssignment);
+            userTaskAssignmentListener = (assignment, subscription) -> {
                 assignor.onAssignmentComputed(assignment, subscription, assignmentError);
             };
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Really doesn't matter, this is just an fyi, but you don't need the brackets for one-line lambdas and we typically use the inline/bracket-less style when possible in Streams code

```suggestion
            userTaskAssignmentListener = (assignment, subscription) -> assignor.onAssignmentComputed(assignment, subscription, assignmentError)
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java [hunk 5]
```diff
@@ -489,10 +489,10 @@ public SortedMap<String, Set<TaskId>> taskIdsByPreviousConsumer() {
...
     public void setAssignedTasks(final KafkaStreamsAssignment assignment) {
-        final Set<TaskId> activeTasks = assignment.assignment().stream()
+        final Set<TaskId> activeTasks = assignment.tasks().values().stream()
             .filter(task -> task.type() == ACTIVE).map(KafkaStreamsAssignment.AssignedTask::id)
             .collect(Collectors.toSet());
-        final Set<TaskId> standbyTasks = assignment.assignment().stream()
+        final Set<TaskId> standbyTasks = assignment.tasks().values().stream()
             .filter(task -> task.type() == STANDBY).map(KafkaStreamsAssignment.AssignedTask::id)
             .collect(Collectors.toSet());
         assignedActiveTasks.taskIds(activeTasks);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java]: nit: I keep getting confused by this and then realizing it's just a poorly-named method, can you rename it to `#setTaskIds` or something like that?

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 6]
```diff
@@ -72,6 +83,27 @@ public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(final Ap
...
+     * @param kafkaStreamsAssignments the current assignment of tasks to KafkaStreams clients
+     *
+     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default standby assignment
+     */
+    public static Map<ProcessId, KafkaStreamsAssignment> defaultStandbyTaskAssignment(final ApplicationState applicationState,
+                                                                                      final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments) {
+        if (!applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty()) {
+            return tagBasedStandbyTaskAssignment(applicationState, kafkaStreamsAssignments);
+        } else if (canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY)) {
+            return tagBasedStandbyTaskAssignment(applicationState, kafkaStreamsAssignments);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:98]: I'm guessing this represents the middle conditional in the `StandbyTaskAssignorFactory#create` method? I was struggling to understand the point of that because how can you do tag-based rack-aware assignment with no rack-aware assignment tags (ie the first condition)

Then I realized it's actually only ever hit when called from testing code. In the real code, the `rackAwareTaskAssignor` parameter is always null and therefore the middle condition (ie `if (rackAwareTaskAssignor != null && rackAwareTaskAssignor.validClientRack())`) is always false

I have to assume this just exists to make it easier to hook into and/or set up from unit tests, but it's definitely bad practice to have conditions in production code that just exist for test cases. That or it's a bug/regression. Anyways, the tl;dr is I'm pretty sure we should remove these two lines and only have the if/else branches

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 7]
```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
         );
         LOG.info("Assignment before standby task optimization has cost {}", initialCost);
 
-        throw new UnsupportedOperationException("Not yet Implemented.");
+        final MoveStandbyTaskPredicate moveablePredicate = getStandbyTaskMovePredicate(applicationState);
+        final BiFunction<KafkaStreamsAssignment, KafkaStreamsAssignment, List<TaskId>> getMovableTasks = (source, destination) -> {
+            return source.tasks().values().stream().filter(task -> task.type() == AssignedTask.Type.STANDBY)
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: nit: put `.filter(task -> task.type() == AssignedTask.Type.STANDBY)` on a separate line

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 8]
```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+                .filter(task -> {
+                    return !destination.tasks().containsKey(task.id());
+                })
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]:
```suggestion
                .filter(task -> !destination.tasks().containsKey(task.id()))
```
Just makes it a bit easier to read when we keep lambdas on one line by getting rid of the brackets, especially when we have a complex structure containing multiple steps with lambdas, nested lambda, etc

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 9]
```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+            round++;
+            for (int i = 0; i < kafkaStreamsAssignments.size(); i++) {
+                final UUID clientId1 = clientIds.get(i);
+                final KafkaStreamsAssignment clientState1 = kafkaStreamsAssignments.get(new ProcessId(clientId1));
+                for (int j = i + 1; j < kafkaStreamsAssignments.size(); j++) {
+                    final UUID clientId2 = clientIds.get(i);
+                    final KafkaStreamsAssignment clientState2 = kafkaStreamsAssignments.get(new ProcessId(clientId2));
+
+                    final String rack1 = clientRacks.get(clientState1.processId().id()).get();
+                    final String rack2 = clientRacks.get(clientState2.processId().id()).get();
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:305]: Future cleanup: we can actually get rid of this clientRacks data structure altogether, and just use the rack ids we already processed and nicely organized in the KafkaStreamsState which is addressable by UUID, ie

```suggestion
final String rack1 = kafkaStreamsStates.get(clientState1.processId().id()).rackId().get();
final String rack2 = kafkaStreamsStates.get(clientState2.processId().id()).rackId().get();
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 10]
```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+                    final List<TaskId> movable2 = getMovableTasks.apply(clientState2, clientState1);
+
+                    // There's no needed to optimize if one is empty because the optimization
+                    // can only swap tasks to keep the client's load balanced
+                    if (movable1.isEmpty() || movable2.isEmpty()) {
+                        continue;
+                    }
+
+                    final List<TaskId> taskIdList = Stream.concat(movable1.stream(),
+                            movable2.stream())
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: I noticed this is messed up in the original code too, but can you fix the indentation/remove the line break (just here, no need to fix it in the RackAwareTaskAssignor too)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 11]
```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+                        .sorted()
+                        .collect(Collectors.toList());
+
+                    final List<UUID> clients = Stream.of(clientId1, clientId2)
+                        .sorted().collect(
+                            Collectors.toList());
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Same here, remove the random line break before `Collectors.toList());`

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 12]
```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+
+                    taskMoved |= graphConstructor.assignTaskFromMinCostFlow(
+                        assignmentGraph.graph,
+                        clientIds,
+                        taskIds,
+                        assignmentsByUuid,
+                        assignmentGraph.taskCountByClient,
+                        assignmentGraph.clientByTask,
+                        (assignment, taskId) -> assignment.assignTask(new AssignedTask(taskId, AssignedTask.Type.ACTIVE)),
+                        (assignment, taskId) -> assignment.removeTask(new AssignedTask(taskId, AssignedTask.Type.ACTIVE)),
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]:
```suggestion
                        (assignment, taskId) -> assignment.assignTask(new AssignedTask(taskId, AssignedTask.Type.STANDBY)),
                        (assignment, taskId) -> assignment.removeTask(new AssignedTask(taskId, AssignedTask.Type.STANDBY)),
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 13]
```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+                        (assignment, taskId) -> assignment.tasks().containsKey(taskId)
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]:
```suggestion
                        (assignment, taskId) -> assignment.tasks().containsKey(taskId) && assignment.tasks().get(taskId).type() == AssignedTask.Type.STANDBY
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 14]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
...
+        final Map<UUID, KafkaStreamsAssignment> clients = kafkaStreamsAssignments.entrySet().stream().collect(Collectors.toMap(
+            entry -> entry.getKey().id(),
+            Map.Entry::getValue
+        ));
+
+        final ConstrainedPrioritySet standbyTaskClientsByTaskLoad = standbyTaskPriorityListByLoad(streamStates, kafkaStreamsAssignments);
+        standbyTaskClientsByTaskLoad.offerAll(streamStates.keySet().stream().map(ProcessId::id).collect(Collectors.toSet()));
+        for (final TaskId task : statefulTaskIds) {
+            assignStandbyTasksForActiveTask(numStandbyReplicas, clients,
+                tasksToRemainingStandbys, standbyTaskClientsByTaskLoad, task);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: nit: another weird linebreak, should be either all parameters on the same line or all on separate lines (though I know this wasn't your doing)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ConstrainedPrioritySet.java [hunk 15]
```diff
@@ -30,13 +30,13 @@
...
-class ConstrainedPrioritySet {
+public class ConstrainedPrioritySet {
 
     private final PriorityQueue<UUID> clientsByTaskLoad;
     private final BiFunction<UUID, TaskId, Boolean> constraint;
     private final Set<UUID> uniqueClients = new HashSet<>();
 
-    ConstrainedPrioritySet(final BiFunction<UUID, TaskId, Boolean> constraint,
+    public ConstrainedPrioritySet(final BiFunction<UUID, TaskId, Boolean> constraint,
                            final Function<UUID, Double> weight) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ConstrainedPrioritySet.java]: nit: adjust parameter indentation

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 16]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
...
+        int numRemainingStandbys = tasksToRemainingStandbys.get(activeTaskId);
+        while (numRemainingStandbys > 0) {
+            final UUID client = standbyTaskClientsByTaskLoad.poll(activeTaskId);
+            if (client == null) {
+                break;
+            }
+            clients.get(client).assignTask(new AssignedTask(activeTaskId, AssignedTask.Type.STANDBY));
+            numRemainingStandbys--;
+            standbyTaskClientsByTaskLoad.offer(client);
+            tasksToRemainingStandbys.put(activeTaskId, numRemainingStandbys);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: nit: kind of confusing to update the map on every iteration of this loop, we should just do it once, when we have the final value after exiting the loop

(I also can't tell if the `tasksToRemainingStandbys` values are ever even accessed again after calling this map, but I'm not 100% sure and it seems best to leave that map in a consistent state just in case it does get used again in the future 🤷‍♀️ )

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 17]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
         }
         return true;
     }
+
+    private static Map<ProcessId, KafkaStreamsAssignment> tagBasedStandbyTaskAssignment(final ApplicationState applicationState,
+                                                                                        final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments) {
+        final int numStandbyReplicas = applicationState.assignmentConfigs().numStandbyReplicas();
+        final Map<TaskId, Integer> tasksToRemainingStandbys = applicationState.allTasks().values().stream()
+            .collect(Collectors.toMap(TaskInfo::id, taskInfo -> numStandbyReplicas));
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: nit: this isn't a correctness issue, but for logging/debugging/inspection sanity we should only fill in this map with stateful tasks. Technically the "remaining standbys" for any stateless tasks is just 0 (though I'd rather just not include them at all in this map, not suggesting you set their value to 0)

We can also save a small bit of processing overhead as well by just building this map based on the `statefulTaskIds` set that's initialized a bit later. We can move this down to after that gets created and then simplify it to just `statefulTaskIds.stream().collect(toMap(t -> t, t -> numStandbyReplicas))`

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 18]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
...
+        }
+
+        return kafkaStreamsAssignments;
+    }
+
+    private static Map<ProcessId, KafkaStreamsAssignment> loadBasedStandbyTaskAssignment(final ApplicationState applicationState,
+                                                                                         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments) {
+        final int numStandbyReplicas = applicationState.assignmentConfigs().numStandbyReplicas();
+        final Map<TaskId, Integer> tasksToRemainingStandbys = applicationState.allTasks().values().stream()
+            .collect(Collectors.toMap(TaskInfo::id, taskInfo -> numStandbyReplicas));
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: ditto here, move down a bit and then we can just initialize it as `statefulTaskIds.stream().collect(toMap(t -> t, t -> numStandbyReplicas))`

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 19]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
...
+        }
+    }
+
+    private static List<String> getRackAwareAssignmentTags(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+
+        if (hasRackAwareAssignmentTags) {
+            return applicationState.assignmentConfigs().rackAwareAssignmentTags();
+        } else if (canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY)) {
+            return Collections.singletonList("rack");
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Similar to [this comment](https://github.com/apache/kafka/pull/16129/files#r1619754391), I don't think the middle case exists in non-testing code, and we shouldn't have special cases for tests in production code.

Also, once you make the simplification from that comment, it should only be possible to reach this if there are rack aware tags. So we should be able to just inline the `applicationState.assignmentConfigs().rackAwareAssignmentTags()` and get rid of this method altogether

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 20]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
...
+        final Map<UUID, KafkaStreamsAssignment> clientsByUuid = kafkaStreamsAssignments.entrySet().stream().collect(Collectors.toMap(
+            entry -> entry.getKey().id(),
+            Map.Entry::getValue
+        ));
+        final Function<KafkaStreamsState, Map<String, String>> clientTagGetter = createClientTagGetter(applicationState);
+
+        final Map<TaskId, ProcessId> pendingStandbyTasksToClientId = new HashMap<>();
+        for (final TaskId statefulTaskId : statefulTaskIds) {
+            for (final KafkaStreamsAssignment assignment : clientsByUuid.values()) {
+                if (assignment.tasks().containsKey(statefulTaskId)) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:563]: one more minor bug: in the original code this is actually checking for assigned _active_ tasks specifically, so this should be 
```suggestion
                if (assignment.tasks().containsKey(statefulTaskId) && assignment.tasks().get(statefulTaskId).type() == Type.ACTIVE) {
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 21]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
...
+            return Collections.singletonList("rack");
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    private static MoveStandbyTaskPredicate getStandbyTaskMovePredicate(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+        final boolean canPerformRackAwareOptimization = canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY);
+        if (hasRackAwareAssignmentTags || canPerformRackAwareOptimization) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Was thinking about this a bit and decided it should just be `if (hasRackAwareAssignmentTags)` -- for one thing, this method should only ever be invoked if `canPerformRackAwareOptimization` is true since it's called like halfway through `#optimizeRackAwareStandbyTasks`.  For another, everything inside this `if` case is all about client tags, so it seems like `hasRackAwareAssignmentTags` must be true to enter this part of the code. 

tl;dr we can remove the `canPerformRackAwareOptimization` variable and method call from this method entirely

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 22]
```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
...
+    }
+
+    private static Function<KafkaStreamsState, Map<String, String>> createClientTagGetter(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+        final boolean canPerformRackAwareOptimization = canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY);
+
+        if (hasRackAwareAssignmentTags || !canPerformRackAwareOptimization) {
+            return KafkaStreamsState::clientTags;
+        } else {
+            return state -> mkMap(mkEntry("rack", state.rackId().get()));
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: IIUC this also relates back to [this whole thing](https://github.com/apache/kafka/pull/16129/files#r1619754391). So if we get rid of that, we should be able to simplify (or even outright remove) this method as well

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:98]: This is driving me crazy, why add so much complexity with these fake "rack" tags if we never use it in non-testing code? I looked up the PR that added this whole thing in the first place ([#14097](https://github.com/apache/kafka/pull/14097)) and the description makes it sound like the "rack" tags are supposed to be a real feature, essentially piggybacking on the older tag-based rack-aware assignment logic. Yet even in that PR, the only thing that actually uses it is from test code. So now I can't tell if it's just an oversight/bug, an unfinished feature that was forgotten about, or really truly just meant to make it easier to write tests. 

Anyways I just need to ask Hao (guy who did the rack-aware optimization KIP) at this point. So in the meantime, until I hear back from him or one of the Confluent folks can answer for him, feel free to ignore all of my comments that relate to this (there are many since this resulted in a lot of branching logic and complexity)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java (hunk 3 above)
Comment (apourchet) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java:281]: Created a ticket for it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 23]
```diff
@@ -555,18 +556,21 @@ private ApplicationState buildApplicationState(final TopologyMetadata topologyMe
...
-                !stateStoreNames.isEmpty(),
-                stateStoreNames,
-                topicPartitions
-            );
-        }).collect(Collectors.toSet());
+        final Map<TaskId, TaskInfo> logicalTasks = logicalTaskIds.stream().collect(Collectors.toMap(
+            Function.identity(),
+            taskId -> {
+                final Set<String> stateStoreNames = topologyMetadata
+                    .stateStoreNameToSourceTopicsForTopology(taskId.topologyName())
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:563]: Ah somehow I missed this before -- this is actually returning _all_ the state stores for this topology, it's not specific to the taskId. This was an existing issue so we don't need to fix it in this PR, it can be addressed in a followup. It might be a bit complicated so I'll take a look at how we can get this info

Would've caught this during testing since we definitely want tests with mixed stateless-and-stateful tasks, but still good to fix ASAP

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 24]
```diff
@@ -72,6 +80,27 @@ public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(final Ap
...
+     * @param kafkaStreamsAssignments the current assignment of tasks to KafkaStreams clients
+     *
+     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default standby assignment
+     */
+    public static Map<ProcessId, KafkaStreamsAssignment> defaultStandbyTaskAssignment(final ApplicationState applicationState,
+                                                                                      final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments) {
+        if (!applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty()) {
+            return tagBasedStandbyTaskAssignment(applicationState, kafkaStreamsAssignments);
+        } else if (canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY)) {
+            return tagBasedStandbyTaskAssignment(applicationState, kafkaStreamsAssignments);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:98]: Address in a followup:

We should just remove this case entirely right? basically it's "if hasRackAwareTags then do tag-based standby task assignment, if doesNotHaveTags then do default standby task assignment"

Note that the tag-based rack aware assignment has nothing to do with the rack ids. So `canPerformRackAwareOptimization` is kind of irrelevant to the question here

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 25]
```diff
@@ -244,18 +271,112 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+                    final AssignmentGraph assignmentGraph = buildTaskGraph(
+                        assignmentsByUuid,
+                        clientRacks,
+                        taskIdList,
+                        clients,
+                        topicPartitionsByTaskId,
+                        crossRackTrafficCost,
+                        nonOverlapCost,
+                        false,
+                        false,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:334]: followup PR: should be true, true I think?

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 20 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:563]: @apourchet make sure to address this in the followup PR

**Review by ableegoldman — COMMENTED** [review]
A few more things that need to be addressed but can be done in a followup PR

**Comment by ableegoldman** [comment]
Test failures are unrelated. Merging to trunk

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 25 above)
Comment (apourchet) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:334]: you're right, good catch!
